scripts/lib/ts-generator.js

```diff
@@ -0,0 +1,437 @@
+        var waiterModelFile = fs.readFileSync(waiterFilePath);
+        var waiterModel = JSON.parse(waiterModelFile);
+        serviceModel.waiters = waiterModel.waiters;
+    }
+
+    return serviceModel;
+};
+
+/**
+ * Determines if a member is required by checking for it in a list.
+ */
+TSGenerator.prototype.checkRequired = function checkRequired(list, member) {
+    if (list.indexOf(member)) {
+        return true;
+    }
+    return false;
+};
+
+/**
+ * Generates whitespace based on the count.
+ */
+TSGenerator.prototype.tabs = function tabs(count) {
+    var code = '';
+    for (var i = 0; i < count; i++) {
+        code += '  ';
+    }
+    return code;
+};
+
+/**
+ * Transforms documentation string to a more readable format.
+ */
+TSGenerator.prototype.transformDocumentation = function transformDocumentation(documentation) {
+    if (!documentation) {
+        return '';
+    }
+    documentation = documentation.replace(/<(?:.|\n)*?>/gm, '');
+    documentation = documentation.replace(/\*\//g, '*');
+    return documentation;
+};
+
+/**
+ * Returns a doc string based on the supplied documentation.
+ * Also tabs the doc string if a count is provided.
+ */
+TSGenerator.prototype.generateDocString = function generateDocString(documentation, tabCount) {
+    tabCount = tabCount || 0;
+    var code = '';
+    code += this.tabs(tabCount) + '/**\n';
+    code += this.tabs(tabCount) + ' * ' + this.transformDocumentation(documentation) + '\n';
+    code += this.tabs(tabCount) + ' */\n';
+    return code;
+};
+
+/**
+ * Returns an array of custom configuration options based on a service identiffier.
+ * Custom configuration options are determined by checking the metadata.json file.
+ */
+TSGenerator.prototype.generateCustomConfigFromMetadata = function generateCustomConfigFromMetadata(serviceIdentifier) {
+    // some services have additional configuration options that are defined in the metadata.json file
+    // i.e. dualstackAvailable = useDualstack
+    // create reference to custom options
+    var customConfigurations = [];
+    var serviceMetadata = this.metadata[serviceIdentifier];
+    // loop through metadata members
+    for (var memberName in serviceMetadata) {
+        if (!serviceMetadata.hasOwnProperty(memberName)) {
+            continue;
+        }
+        var memberValue = serviceMetadata[memberName];
+        // check configs
+        switch (memberName) {
+            case 'dualstackAvailable':
+                customConfigurations.push(CUSTOM_CONFIG_ENUMS.DUALSTACK);
+                break;
+        }
+    }
+
+    return customConfigurations;
+};
+
+/**
+ * Generates a type or interface based on the shape.
+ */
+TSGenerator.prototype.generateTypingsFromShape = function generateTypingsFromShape(shapeKey, shape, tabCount) {
+    // some shapes shouldn't be generated if they are javascript primitives
+    if (['string', 'boolean', 'number', 'Date', 'Blob'].indexOf(shapeKey) >= 0) {
+        return '';
```

Comment:
So for example, the typing could look like:
```
export type _Date = string;
```
And the references to this shape would have to have an underscore added in front as well.